src/Catalog/PackageCatalogItemCreator.cs

```diff
@@ -0,0 +1,240 @@
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NuGet.Services.Metadata.Catalog.Helpers;
+using NuGet.Services.Metadata.Catalog.Persistence;
+
+namespace NuGet.Services.Metadata.Catalog
+{
+    public sealed class PackageCatalogItemCreator
+    {
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
+        private readonly IAzureStorage _storage;
+        private readonly ITelemetryService _telemetryService;
+
+        private PackageCatalogItemCreator(
+            HttpClient httpClient,
+            ITelemetryService telemetryService,
+            ILogger logger,
+            IStorage storage)
+        {
+            _httpClient = httpClient;
+            _telemetryService = telemetryService;
+            _logger = logger;
+            _storage = storage as IAzureStorage;
+        }
+
+        public static PackageCatalogItemCreator Create(
+            HttpClient httpClient,
+            ITelemetryService telemetryService,
+            ILogger logger,
+            IStorage storage)
+        {
+            if (httpClient == null)
+            {
+                throw new ArgumentNullException(nameof(httpClient));
+            }
+
+            if (telemetryService == null)
+            {
+                throw new ArgumentNullException(nameof(telemetryService));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            return new PackageCatalogItemCreator(httpClient, telemetryService, logger, storage);
+        }
+
+        public async Task<PackageCatalogItem> CreateAsync(
+            FeedPackageDetails packageItem,
+            DateTime timestamp,
+            CancellationToken cancellationToken)
+        {
+            if (packageItem == null)
+            {
+                throw new ArgumentNullException(nameof(packageItem));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            PackageCatalogItem item = null;
+
+            if (_storage != null)
+            {
+                item = await GetPackageViaStorageAsync(packageItem, cancellationToken);
+            }
+
+            if (item == null)
+            {
+                item = await GetPackageViaHttpAsync(packageItem, timestamp, item, cancellationToken);
+            }
+
+            return item;
+        }
+
+        private async Task<PackageCatalogItem> GetPackageViaStorageAsync(FeedPackageDetails packageItem, CancellationToken cancellationToken)
+        {
+            PackageCatalogItem item = null;
+            var packageFileName = PackageUtility.GetPackageFileNameLowercase(packageItem.PackageId, packageItem.PackageVersion);
+            var blob = _storage.GetCloudBlockBlobReference(packageFileName);
+
+            if (blob == null)
+            {
```

Comment:
Minor suggestion: I think this would read slightly better if this condition returned null. Then, the rest of this method wouldn't need to be nested in an `else` block.